merkle-tree/indexed/src/lib.rs

```diff
@@ -179,11 +186,28 @@ where
     pub fn update(
         &mut self,
         changelog_index: usize,
-        new_element: IndexedElement<I>,
+        mut new_element: IndexedElement<I>,
         low_element: IndexedElement<I>,
         low_element_next_value: &BigUint,
         low_leaf_proof: &mut BoundedVec<[u8; 32]>,
     ) -> Result<(), IndexedMerkleTreeError> {
+        // Check whether the `next_index` of `low_element` was changed in the
+        // changelog. If yes, we need to use that new value in
+        // `new_element.next_index`.
+        if let Some(current_low_element_next_index) = self.changelog.get(&low_element.value)? {
+            // If `low_element.next_index` is up to date, don't do anything. At
+            // this point any mistakes leading to errors are responsity of the
+            // caller.
+            if &low_element.next_index != current_low_element_next_index {
+                // There are two cases.
+                // 1. The new `current_low_element_index` points to the value
+                //    lower than `new_element`. In such case, that element
+                //    becomes a new low element.
+                // 2. It points to the value higher than `new_element`. In such
+                //    case, it becomes the next element of `new_element`.
+            }
+        }
```

Comment:
I think this is the most important piece of the address Merkle tree concurrency.

What do you think about putting this into a separate function that we call prior to calling the update method.
That way it's more modular and doesn't mess with any existing code and tests.

Reply:
Sounds good, I can do that.